### PR TITLE
split core::ptr module into multiple files

### PR DESCRIPTION
**Pull request by RalfJung**
Cc @Centril

### CODE REVIEW

File: src/libcore/ptr/non_null.rs [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    fn hash<H: hash::Hasher>(&self, state: &mut H) {
+        self.as_ptr().hash(state)
+    }
+}
+
+#[unstable(feature = "ptr_internals", issue = "0")]
+impl<T: ?Sized> From<Unique<T>> for NonNull<T> {
+    #[inline]
+    fn from(unique: Unique<T>) -> Self {
+        unsafe { NonNull::new_unchecked(unique.as_ptr()) }
```
Comment (RalfJung) [inline src/libcore/ptr/non_null.rs]: I had to rewrite this using the public API of `Unique`.

File: src/libcore/ptr/unique.rs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    pub fn new(ptr: *mut T) -> Option<Self> {
+        if !ptr.is_null() {
+            Some(unsafe { Unique { pointer: ptr as _, _marker: PhantomData } })
+        } else {
+            None
+        }
+    }
+
+    /// Acquires the underlying `*mut` pointer.
+    #[inline]
```
Comment (RalfJung) [inline src/libcore/ptr/unique.rs]: I made this and the other methods here `inline` for consistency with `NonNull`.

File: src/libcore/ptr/unique.rs [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    fn from(reference: &T) -> Self {
+        unsafe { Unique { pointer: reference as *const T, _marker: PhantomData } }
+    }
+}
+
+#[unstable(feature = "ptr_internals", issue = "0")]
+impl<'a, T: ?Sized> From<NonNull<T>> for Unique<T> {
+    #[inline]
+    fn from(p: NonNull<T>) -> Self {
+        unsafe { Unique::new_unchecked(p.as_ptr()) }
```
Comment (RalfJung) [inline src/libcore/ptr/unique.rs]: I had to adjust this to use the public API of `NonNull`.

**Comment by clarfon** [comment]
Glad to see more of these PRs since #42523 and #56932 :p

**Pull request merged**